File: sandbox/demo/diag.py
import logging
import os

from sandbox.demo.fs import read_text
from sandbox.demo.utils import select_lines_with_prefix

# proc status keys to log in debug mode
STATUS_KEYS: tuple[str, ...] = ("Uid:", "Gid:", "Groups:", "CapEff:", "NSpid:")


def log_identity(
    logger: logging.Logger, label: str, host_uid: int, host_gid: int
) -> None:
    uid, euid = os.getuid(), os.geteuid()
    gid, egid = os.getgid(), os.getegid()

    ruid, reuid, suid = os.getresuid()
    rgid, regid, sgid = os.getresgid()

    logger.debug(f"=== {label} ===")
    logger.debug(f"host: uid={host_uid},gid={host_gid}")
    logger.debug(f"self: uid={uid},euid={euid},gid={gid},egid={egid}")
    logger.debug(f"self: resuid={ruid},{reuid},{suid} resgid={rgid},{regid},{sgid}")

    uid_map = read_text("/proc/self/uid_map")
    gid_map = read_text("/proc/self/gid_map")
    logger.debug(f"/proc/self/uid_map:\n{uid_map}")
    logger.debug(f"/proc/self/gid_map:\n{gid_map}")

    status = read_text("/proc/self/status")
    for line in select_lines_with_prefix(status, STATUS_KEYS):
        logger.debug(line)

File: sandbox/sandbox/diag/diag.go
package diag

import (
	"errors"
	"fmt"
	"log"
	"os"
	"syscall"

	"sandbox/sandbox/utils"
)

var StatusKeys = []string{"Uid:", "Gid:", "Groups:", "CapEff:", "NSpid:"}

func LogIdentity(label string, hostUID, hostGID int) {
	uid := syscall.Getuid()
	euid := syscall.Geteuid()
	gid := syscall.Getgid()
	egid := syscall.Getegid()

	log.Printf("[DEBUG] === %s ===", label)
	log.Printf("[DEBUG] host: uid=%d,gid=%d", hostUID, hostGID)
	log.Printf("[DEBUG] self: uid=%d,euid=%d,gid=%d,egid=%d", uid, euid, gid, egid)

	uidMap := ReadText("/proc/self/uid_map")
	gidMap := ReadText("/proc/self/gid_map")
	log.Printf("[DEBUG] /proc/self/uid_map:\n%s", uidMap)
	log.Printf("[DEBUG] /proc/self/gid_map:\n%s", gidMap)

	statusText := ReadText("/proc/self/status")
	matchedLines := utils.SelectLinesWithPrefix(statusText, StatusKeys)
	for _, line := range matchedLines {
		log.Printf("[DEBUG] %s", line)
	}
}

func ReadText(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return "<missing>"
		}
		if errors.Is(err, os.ErrPermission) {
			return fmt.Sprintf("<permission denied: %v>", err)
		}
		return fmt.Sprintf("<error: %v>", err)
	}
	return string(data)
}
